opt_einsum/paths.py

```diff
@@ -88,6 +91,165 @@ def optimal(input_sets, output_set, idx_dict, memory_limit):
     return path
 
 
+def _calc_k12_flops(inputs, output, remaining, i, j, size_dict):
+    k1, k2 = inputs[i], inputs[j]
+    either = k1 | k2
+    shared = k1 & k2
+
+    other_inputs = tuple(inputs[p] for p in remaining - {i, j})
+    keep = frozenset.union(output, *other_inputs)
+
+    cost = helpers.flop_count(either, shared - keep, 2, size_dict)
+
+    k12 = either & keep
+    return k12, cost
+
+
+def roptimal(inputs, output, size_dict, memory_limit=None):
+
+    inputs = tuple(map(frozenset, inputs))
+    output = frozenset(output)
+
+    best = {'flops': inf, 'path': (tuple(range(len(inputs))),)}
+    size_cache = {}
+    result_cache = {}
+
+    def _optimal_iterate(path, remaining, inputs, flops):
+
+        # reached end of path (only ever get here if flops is best found so far)
+        if len(remaining) == 1:
+            best['flops'] = flops
+            best['path'] = path
+
+        # check all possible remaining paths
+        for i, j in itertools.combinations(remaining, 2):
+            k1, k2 = inputs[i], inputs[j]
+            try:
+                k12, flops12 = result_cache[k1, k2]
+            except KeyError:
+                k12, flops12 = result_cache[k1, k2] = _calc_k12_flops(inputs, output, remaining, i, j, size_dict)
+
+            # sieve based on current best flops
+            new_flops = flops + flops12
+            if new_flops >= best['flops']:
+                continue
+
+            # sieve based on memory limit
+            if memory_limit is not None:
+                try:
+                    size12 = size_cache[k12]
+                except KeyError:
+                    size12 = size_cache[k12] = helpers.compute_size_by_dict(k12, size_dict)
+
+                if size12 > memory_limit:
+                    continue
+
+            _optimal_iterate(path=path + ((i, j),),
+                             inputs=inputs + (k12,),
+                             remaining=remaining - {i, j} | {len(inputs)},
+                             flops=new_flops)
+
+    _optimal_iterate(path=(),
+                     inputs=inputs,
+                     remaining=set(range(len(inputs))),
+                     flops=0)
+
+    return ssa_to_linear(best['path'])
+
+
+def rgreedy(inputs, output, size_dict, memory_limit=None, nbranch=None):
+
+    inputs = tuple(map(frozenset, inputs))
+    output = frozenset(output)
+
+    best = {'flops': inf, 'path': (tuple(range(len(inputs))),)}
+    best_progress = defaultdict(lambda: inf)
+
+    size_cache = {k: helpers.compute_size_by_dict(k, size_dict) for k in inputs}
+    result_cache = {}
+
+    def _greedy_iterate(path, inputs, remaining, flops):
+
+        # reached end of path (only ever get here if flops is best found so far)
+        if len(remaining) == 1:
+            best['flops'] = flops
+            best['path'] = path
+
+        def _assess_candidate(k1, k2, i, j):
+            # find resulting indices and cost
+            try:
```

Comment:
Can we do a `if (k1, k2) in result_cache` here and below? Although I guess it might be slower... maybe:

```
k12, flops12 = result_cache.setdefault((k1, k2), _calc_k12_flops(inputs, output, remaining, i, j, size_dict))
```

Reply:
This is, marginally, the quickest I've found! I think because the key is more often in the cache than not - otherwise checking membership first would edge it in terms of performance. The ``setdefault`` is definitely elegant but I think it has to evaluate the ``default`` arg everytime, even if ``k1, k2`` is in the cache.

Reply:
Maybe `value = cache.get(key) or cache.setdefault(key, func(...))`?

Reply:
I think my comment above applies still, the ``try...except`` with standard ``__getitem__``/``__setitem__`` syntax performs best overall. The way that python works means ``func(...)`` will be eagerly called and supplied to ``setdefault`` even if it is not needed sadly.

Reply:
From running a sample ``optimal`` the timings I get are:
```
try-exc / setdefault : 127 ms ± 2.26 ms
try-exc /   setitem  : 131 ms ± 4.18 ms
 not-in /   setitem  : 141 ms ± 1.33 ms
 not-in / setdefault : 148 ms ± 3.86 ms
  get   /   setitem  : 146 ms ± 1.97 ms
  get   / setdefault : 149 ms ± 2.83 ms
```
so there does seem to be a 5-10% advantage - I think just because the cache is hit more often than missed. However, these timings are marginal (maybe machine dependant etc), and though I have always thought of ``try/except`` as 'pythonic', completely happy to defer to whichever you want as long as you don't mind a possible slight performance hit!

Reply:
Interesting do you precompute the key in your timings `key = str((k1, k2))`?

Either way, can you tuck this functionality into a util function so that we can revisit later?

Reply:
In my experience, the two standard patterns are: `try...except` if keys are usually present, and `if ... in ... else` if keys are usually absent. `setdefault` is only used if value computation is free.

Reply:
> Interesting do you precompute the key in your timings key = str((k1, k2))?

I'm using ``key = (k1, k2)`` now (marginal saving over forming the tuple twice), this is already hashable due to being frozensets so applying ``str`` or ``hash`` doesn't help I find. 

> Kind of interesting that you get this setdefault/setitem reversal in your timings.

I suspect this is possibly just noise? I'd be happy to use either but I fell like the standard setitem is simpler here as we know the key is not in the cache already.

> Either way, can you tuck this functionality into a util function so that we can revisit later?

So, I did try this with a ``lazy_setdefault`` function, but you get quite a slow down - these few lines are a real hot spot where these micro optimizations make a big difference it seems!

Overall I'm leaning towards leaving it roughly as it is?

```python
key = (inputs[i], inputs[j])
try:
    k12, flops12 = result_cache[key]
except KeyError:
    k12, flops12 = result_cache[key] = _calc_k12_flops(inputs, output, remaining, i, j, size_dict)
```

Reply:
Alright, at some point I am just complaining for no good reason. If you could cast it in a function that did the try/except/lazy evaluate that would be best, but no reason to eat the overhead if it costs too much.